Fix shifting logical operators up in filter parsing

Nitwel opened the pull request:
## Scope

What's changed:
- the shiftLogicalOperatorsUp function now does what it's name said. In that before, each filter only ever got shifted once up and not all the way to the top. 

## Potential Risks / Drawbacks

- None

## Review Notes / Questions

The problem was that the GQL query in the PR generated a filter that had a logical operator 2 layers deep.

- See tests for examples

---

Fixes #24367

Nitwel committed:
fix carrying up logical op in filter

licitdev committed:
Update changeset

licitdev committed:
Add nested logical operators test

licitdev commented on packages/utils/shared/parse-filter.test.ts:
```diff
@@ -67,6 +67,163 @@ describe('#parseFilter', () => {
+			},
+		} as Filter;
+
+		const mockResult = {
+			_or: [
+				{
+					_and: [
+						{
+							i_really_dont_know: {
+								article_field: { date_field: { _gte: '2023-10-01T00:00:00' } },
+							},
+						},
+						{
+							i_really_dont_know: {
+								article_field: { date_field: { _lt: '2023-11-01T00:00:00' } },
+							},
+						},
+					],
+				},
+				{
+					_and: [
+						{
+							i_really_dont_know: {
+								other_field: { date_field: { _gte: '2023-10-01T00:00:00' } },
+							},
+						},
+						{
+							i_really_dont_know: {
+								other_field: { date_field: { _lt: '2023-11-01T00:00:00' } },
+							},
+						},
+					],
+				},
+			],
+		} as Filter;
```
Added a new failing test. Is the expected mock result accurate?

Nitwel replied:
Looks accurate, nice catch! I'll look into fixing that.

licitdev committed:
Try to fix the nested case

Nitwel committed:
add more tests and clean up code

Nitwel commented:
Added 2 more tests and cleaned up the code a bit. Other than that I couldn't think of any other cases that would be worth testing for and the logic also looks solid.
